### DISCUSSION TITLE
Will v2 support NextJS?

### DISCUSSION THREAD

**Discussion by pm-dev** [MichalLytek/type-graphql#1687] [Q&A]
Just migrated my project to NextJS, but doesn't look like the type-graphql reflection works.
```
 ✓ Linting and checking validity of types    
   Collecting page data  ..iv [Error]: Unable to infer GraphQL type from TypeScript reflection system. You need to provide explicit type for 'contains' of 's_' class.
    at iP (/Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:30:3245)
    at /Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:30:10644
    at g (/Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:230:1654)
    at Reflect.<anonymous> (/Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:230:11433)
    at a (/Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:467:37626)
    at 87339 (/Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:36:773)
    at t (/Users/peter/exampleapp/web/.next/server/webpack-runtime.js:1:143)
    at /Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:1:2878
    at t.a (/Users/peter/exampleapp/web/.next/server/webpack-runtime.js:1:908)
    at 11487 (/Users/peter/exampleapp/web/.next/server/app/api/graphql/route.js:1:2791)
```

I noticed https://github.com/MichalLytek/type-graphql/issues/55 so looks like this may be expected. Is there expected support for NextJS in the future or a known workaround?

**Comment by MichalLytek** [comment]
In a company where I work we have TypeGraphQL, TypeScript and NextJS without any issues, so I think it's doable but need to check what config or packages were used to make it work properly.

**Reply by pm-dev** [reply]
Thanks. Would be curious to see this set up with nextjs14. I saw a [tutorial](https://mahieyin-rahmun.medium.com/how-to-setup-a-graphql-server-in-next-js-with-api-endpoints-27aa4001c20b) which uses type-graphql with nextjs, but it uses babel. By default nextjs uses SWC which [may or may not support](https://github.com/swc-project/swc/discussions/5053#discussioncomment-4990976) type-graphql

**Reply by eduardolundgren** [reply]
@MichalLytek Is your company using TypeGraphQL with Next App Router and SWC instead of Babel? I've been trying to get TypeGraphQL to work with Next 14 or 15, but it isn't working for me, even when I use Babel on Next. I am still trying to figure it out and will share here if I find anything that works. If anyone has figured this out, please share.

**Reply by eduardolundgren** [reply]
After more testing using Next 14 building with Babel and SWC, the issue is consistently this one: `ReferenceError: Cannot access 'DocumentCreateNestedOneWithoutKnowledgeInput' before initialization`.

Next requires `"isolatedModules": true` for SWC or Babel, and this seems to be incompatible with TypeGraphQL, which requires cyclic dependencies in the generated code. With isolatedModules, it won't work, as explained in [this message](https://github.com/MichalLytek/typegraphql-prisma/issues/2#issuecomment-744235707) from @MichalLytek.

@MichalLytek, could we modify the generated code to make it compatible with isolatedModules? Do you have any advice for working around this issue? This is important as we are trying to standardize a project on Next 14.


```
> @docana/platform@1.0.0 dev
> next dev -p 8084

  ▲ Next.js 14.2.3
  - Local:        http://localhost:8084
  - Environments: .env.development, .env

 ✓ Starting...
 ✓ Ready in 1074ms
 ○ Compiling /api/graphql ...
 ✓ Compiled /api/graphql in 8.9s (4929 modules)
 ⨯ ReferenceError: Cannot access 'DocumentCreateNestedOneWithoutKnowledgeInput' before initialization
    at Module.DocumentCreateNestedOneWithoutKnowledgeInput (webpack-internal:///(rsc)/./src/server/graphql/generated/resolvers/inputs/DocumentCreateNestedOneWithoutKnowledgeInput.ts:3:91)
    at eval (webpack-internal:///(rsc)/./src/server/graphql/generated/resolvers/inputs/KnowledgeCreateWithoutUserInput.ts:88:178)
    ```

**Reply by eduardolundgren** [reply]
The workaround suggested in [this thread](https://github.com/MichalLytek/typegraphql-prisma/issues/2#issuecomment-1756658656) helped. It recommends omitting all fields that were causing the cyclic dependency, which solved the blocking ReferenceError when using it with Next 14.

`/// @TypeGraphQL.omit(output: true, input: true)`

This workaround is not ideal because it removes those fields from the API. I'm not sure if there's a better approach, but I will keep trying to find another solution and will report back here if I do.
